sheets/api.py

```diff
@@ -0,0 +1,408 @@
+        raise ImproperlyConfigured("Local token file credentials are invalid")
+    return creds
+
+
+def get_credentials():
+    """
+    Gets valid Google API client credentials
+
+    Returns:
+        google.oauth2.credentials.Credentials: Credentials to be used by the Google Drive client/pygsheets/etc.
+
+    Raises:
+        ImproperlyConfigured: Raised if no credentials have been configured
+    """
+    google_api_auth = GoogleApiAuth.objects.order_by("-updated_on").first()
+    if google_api_auth:
+        return Credentials(
+            token=google_api_auth.access_token,
+            refresh_token=google_api_auth.refresh_token,
+            token_uri=GOOGLE_TOKEN_URI,
+            client_id=settings.DRIVE_CLIENT_ID,
+            client_secret=settings.DRIVE_CLIENT_SECRET,
+            scopes=REQUIRED_GOOGLE_API_SCOPES,
+        )
+    if settings.ENVIRONMENT == "dev":
+        token_file_path = os.path.join(settings.BASE_DIR, DEV_TOKEN_PATH)
+        if os.path.exists(token_file_path):
+            return get_google_creds_from_pickled_token_file(token_file_path)
+    raise ImproperlyConfigured("Authorization with Google has not been completed.")
+
+
+def get_data_rows(worksheet):
+    """
+    Yields the data rows of a spreadsheet that has a header row
+
+    Args:
+        worksheet (pygsheets.worksheet.Worksheet): Worksheet object
+
+    Yields:
+        list of str: List of cell values in a given row
+    """
+    row_iter = iter(
+        worksheet.get_all_values(
+            include_tailing_empty=False, include_tailing_empty_rows=False
+        )
+    )
+    # Skip header row
+    next(row_iter)
+    yield from row_iter
+
+
+def create_coupons_for_request_row(coupon_req_row):
+    """
+    Creates coupons for a given request
+
+    Args:
+        coupon_req_row (CouponRequestRow): A representation of a coupon request row
+
+    Returns:
+        CouponGenerationRequest or None: The record of the completed coupon generation request, or None
+            if the request for the given transaction id has already been completed.
+    """
+    with transaction.atomic():
+        coupon_gen_request, created = CouponGenerationRequest.objects.select_for_update().get_or_create(
+            transaction_id=coupon_req_row.transaction_id
+        )
+        if not created and coupon_gen_request.completed:
+            log.error(
+                "Found completed CouponGenerationRequest, but the 'processed' column "
+                "in the spreadsheet == False (transaction id: %s)"
+                % coupon_req_row.transaction_id
+            )
+            return
+        create_coupons(
+            name=coupon_req_row.coupon_name,
+            product_ids=coupon_req_row.get_product_ids(),
+            num_coupon_codes=coupon_req_row.num_codes,
+            coupon_type=CouponPaymentVersion.SINGLE_USE,
+            max_redemptions=coupon_req_row.num_codes,
+            company_id=Company.objects.get(name=coupon_req_row.company_name).id,
```

Comment:
This will cause an exception if the sheet contains a company name that differs slightly from what's in the database, not sure if that's worth worrying about.